llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/reflective/tool_interactive_reflection.py

```diff
@@ -0,0 +1,417 @@
+            )
+            new_steps = []
+        else:
+            correct_msg = self._correct(
+                input_str=input_str, critique=critique_response.response
+            )
+            agent_response = AgentChatResponse(
+                response=str(correct_msg), sources=critique_response.sources
+            )
+            task.extra_state["new_memory"].put(correct_msg)
+
+            if self.max_iterations == state["count"]:
+                new_steps = []
+            else:
+                new_steps = [
+                    step.get_next_step(
+                        step_id=str(uuid.uuid4()),
+                        # NOTE: input is unused
+                        input=None,
+                        step_state=state,
+                    )
+                ]
+
+        return TaskStepOutput(
+            output=agent_response,
+            task_step=step,
+            is_last=is_done | (self.max_iterations == state["count"]),
+            next_steps=new_steps,
+        )
+
+    # Async Methods
+    @dispatcher.span
+    async def _acritique(self, input_str: str) -> AgentChatResponse:
+        agent = self._critique_agent_worker.as_agent(verbose=self._verbose)
+        critique = await agent.achat(
+            self._critique_template.format(input_str=input_str)
+        )
+        if self._verbose:
+            print(f"Critique: {critique.response}", flush=True)
+        return critique
+
+    @dispatcher.span
+    async def _acorrect(self, input_str: str, critique: str) -> ChatMessage:
+        correction = await self._correction_program.acall(
+            input_str=input_str, critique=critique
+        )
+
+        correct_response_str = CORRECT_RESPONSE_FSTRING.format(
+            correction=correction.correction
+        )
+        if self._verbose:
+            print(f"Correction: {correction.correction}", flush=True)
+        return ChatMessage.from_str(correct_response_str, role="assistant")
+
+    @dispatcher.span
+    @trace_method("run_step")
+    async def arun_step(
+        self, step: TaskStep, task: Task, **kwargs: Any
+    ) -> TaskStepOutput:
+        """Run step (async)."""
+        state = step.step_state
+        state["count"] += 1
+
+        messages = task.extra_state["new_memory"].get()
+        current_response = messages[-1].content
+
+        # critique
+        input_str = current_response.replace(
+            "Here is a corrected version of the input.\n", ""
+        )
+        critique_response = await self._acritique(input_str=input_str)
+        task.extra_state["sources"].extend(critique_response.sources)
+
+        if self.stopping_callable:
+            is_done = self.stopping_callable(critique_str=critique_response.response)
+
+        critique_msg = ChatMessage(
+            role=MessageRole.USER, content=critique_response.response
+        )
+        task.extra_state["new_memory"].put(critique_msg)
+
+        # correct
+        if is_done:
```

Comment:
If `self.stopping_callable` is none, is this just never done?

Reply:
oh in that case it will stop after `max_iterations` has been reached. That gets implemented when building the `TaskStepOutput` with `is_last` field.

```python
        return TaskStepOutput(
            output=agent_response,
            task_step=step,
            is_last=is_done | (self.max_iterations == state["count"]),
            next_steps=new_steps,
        )
```

Reply:
Hmm, interesting. I wonder if that's ideal or not. I would expect a combination of the main agent being is_done and the reflection agent being is_done to control when we return 🤔 Maybe work for a future PR

Reply:
oh interesting yea, at this point it always go to reflection and this `is_done` is about when reflection/correction phases stop.